### PR TITLE
Refactor `rz_core_print_disasm_json`

### PR DESCRIPTION
**Pull request by b1llow**
# SQUASH ME

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Rewrite `rz_core_print_disasm_json` and `core_analysis_bytes_standard` with `rz_core_analysis_bytes`

Partially refactored `rz_core_print_disasm`

Remove aliases(associate: https://github.com/rizinorg/rizin/pull/2552):

- `pid` | `pdi` -> `pdq`
- `pId` | `pDi` -> `pDq`
- `pdeqq` -> `pdeQ`
- `pij` -> `pdj`
- `pIj` -> `pDj`

Remove:

- `pd--`
- `pDJ`

In fact, all `pD` and `pd` old code was removed

**Test plan**

...

**Closing issues**

partial https://github.com/rizinorg/rizin/issues/2554

### CODE REVIEW

File: librz/include/rz_core.h [hunk 1]
```diff
@@ -826,7 +837,8 @@ RZ_API RzList *rz_core_asm_bwdisassemble(RzCore *core, ut64 addr, int n, int len
 RZ_API RzList *rz_core_asm_back_disassemble_instr(RzCore *core, ut64 addr, int len, ut32 hit_count, ut32 extra_padding);
 RZ_API RzList *rz_core_asm_back_disassemble_byte(RzCore *core, ut64 addr, int len, ut32 hit_count, ut32 extra_padding);
 RZ_API ut32 rz_core_asm_bwdis_len(RzCore *core, int *len, ut64 *start_addr, ut32 l);
-RZ_API int rz_core_print_disasm(RzPrint *p, RzCore *core, ut64 addr, ut8 *buf, int len, int lines, int invbreak, int nbytes, bool json, PJ *pj, RzAnalysisFunction *pdf);
+RZ_API int rz_core_print_disasm(RZ_NONNULL RzCore *core, ut64 addr, RZ_NONNULL ut8 *buf, int len, int nlines,
+	int invbreak, int cbytes, bool json, RZ_NULLABLE PJ *pj, RZ_NULLABLE RzAnalysisFunction *pdf, RZ_OUT RZ_NULLABLE RzList *out_list);
```
Comment (wargio) [inline librz/include/rz_core.h]: Isn't better to use a structure with the config?

Comment (XVilka) [reply librz/include/rz_core.h]: Agree. Like we made it for strings search function: https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_util/rz_str_search.h#L27

File: librz/core/canalysis.c [hunk 2]
```diff
@@ -6826,41 +6827,66 @@ RZ_API RZ_OWN RzPVector *rz_core_analysis_bytes(RZ_NONNULL RzCore *core, RZ_NONN
...
+	bool asm_sub_var = rz_config_get_i(core->config, "asm.sub.var");
+	core->parser->subrel = rz_config_get_i(core->config, "asm.sub.rel");
+	core->parser->localvar_only = rz_config_get_i(core->config, "asm.sub.varonly");
+
+	const int addrbytes = (int)core->io->addrbytes;
+	const int mask = RZ_ANALYSIS_OP_MASK_ESIL | RZ_ANALYSIS_OP_MASK_IL | RZ_ANALYSIS_OP_MASK_OPEX | RZ_ANALYSIS_OP_MASK_HINT;
+	RzAsmOp asmop;
+	int oplen = 0;
+	char disasm[512];
+	for (int i_ops = 0, i_offset = 0, i_delta = 0; rz_disasm_check_end(nops, i_ops, len, i_delta * addrbytes); i_ops++, i_offset += oplen, i_delta += oplen) {
```
Comment (XVilka) [inline librz/core/canalysis.c]: Better split this line

File: librz/core/disasm.c [hunk 3]
```diff
@@ -303,6 +303,8 @@ typedef struct {
 	int asm_types;
 } RzDisasmState;
 
+static RzList *disasm_texts;
```
Comment (XVilka) [inline librz/core/disasm.c]: Please don't use globals, it will cause huge issues when using in multithreaded  environment

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good overall but there are still problems that should be addressed

File: test/db/tools/rz [hunk 4]
```diff
@@ -1,222 +1 @@
-NAME=rz =
```
Comment (XVilka) [inline test/db/tools/rz]: This is looks like a mistake to remove this file

File: librz/core/cmd/cmd_print.c [hunk 5]
```diff
@@ -7246,73 +6625,57 @@ static void disassembly_n_instructions_as_table(RzTable *t, RzCore *core, int n_
...
-			rz_core_block_size(core, new_blocksize);
-		}
-		rz_core_seek(core, new_offset, true);
-	} else {
-		rz_core_block_read(core);
+	if (!rz_core_handle_backwards_disasm(core, &n_instrs, &n_bytes)) {
+		return false;
 	}
 
+	RZ_LOG_VERBOSE("disassembly at: 0x%" PFMT64x ", blocksize: %" PFMT32d " n_bytes: %" PFMT32d ", n_instrs: %" PFMT32d "\n", core->offset, core->blocksize, n_bytes, n_instrs);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c]: Split this long string please

File: librz/core/disasm.c [hunk 6]
```diff
@@ -301,6 +301,8 @@ typedef struct {
 	const char *strip;
 	int maxflags;
 	int asm_types;
+
+	RzList *list;
```
Comment (XVilka) [inline librz/core/disasm.c]: From what I see in the code we don't really use random insertion and removal of this data, we could use Rz[P]Vector to improve the performance of disassembly greatly, but that's out of the scope of this PR.

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall looks good

**Comment by XVilka** [comment]
@ret2libc please take a look in regard of the command and API changes.

File: librz/core/disasm.c [hunk 7]
```diff
@@ -5274,9 +5319,38 @@ static void ds_end_line_highlight(RzDisasmState *ds) {
 	}
 }
 
-// int l is for lines
-RZ_API int rz_core_print_disasm(RzPrint *p, RzCore *core, ut64 addr, ut8 *buf, int len, int l, int invbreak, int cbytes, bool json, PJ *pj, RzAnalysisFunction *pdf) {
-	int continueoninvbreak = (len == l) && invbreak;
+/**
+ * \brief Free RzAnalysisDisasmText \p p
+ */
+RZ_API void rz_analysis_disasm_text_free(RZ_NULLABLE void *p) {
```
Comment (ret2libc) [inline librz/core/disasm.c]:
```suggestion
RZ_API void rz_analysis_disasm_text_free(RzAnalysisDisasmText *p) {
```

i think we can save ourselves the RZ_NULLABLE in the free cases... It is common practice to allow NULL in all free functions.

Comment (ret2libc) [inline librz/core/disasm.c]: When is this function called?

File: librz/core/disasm.c [hunk 8]
```diff
@@ -5292,17 +5366,19 @@ RZ_API int rz_core_print_disasm(RzPrint *p, RzCore *core, ut64 addr, ut8 *buf, i
...
+	ds->l = nlines;
 	ds->buf = buf;
 	ds->len = len;
 	ds->addr = addr;
 	ds->hint = NULL;
 	ds->buf_line_begin = 0;
-	ds->pdf = pdf;
+	ds->pdf = options ? options->function : NULL;
+	ds->pj = NULL;
+	ds->list = out_list;
```
Comment (ret2libc) [inline librz/core/disasm.c]: why are we getting the list from outside? Shouldn't it be initialized here? e.g. rz_list_newf(rz_disassemble_text_free) ?

File: librz/core/disasm.c [hunk 9]
```diff
@@ -5292,17 +5366,19 @@ RZ_API int rz_core_print_disasm(RzPrint *p, RzCore *core, ut64 addr, ut8 *buf, i
...
 
-	if (json) {
+	if (!out_list && json) {
```
Comment (ret2libc) [inline librz/core/disasm.c]: use ds->list everywhere instead of out_list.

File: librz/include/rz_core.h [hunk 10]
```diff
@@ -836,8 +859,12 @@ RZ_API int rz_core_disasm_pdi(RzCore *core, int nb_opcodes, int nb_bytes, int fm
 RZ_API int rz_core_disasm_pde(RzCore *core, int nb_opcodes, int mode);
 RZ_API RZ_OWN char *rz_core_disasm_instruction(RzCore *core, ut64 addr, ut64 reladdr, RZ_NULLABLE RzAnalysisFunction *fcn, bool color);
 RZ_API bool rz_core_print_function_disasm_json(RzCore *core, RzAnalysisFunction *fcn, PJ *pj);
+RZ_API void rz_core_asm_bb_middle(RZ_NONNULL RzCore *core, ut64 at, RZ_INOUT RZ_NONNULL int *oplen, RZ_NONNULL int *ret);
```
Comment (ret2libc) [inline librz/include/rz_core.h]: is this part of a separate fix? I would move that in a separate PR as this is already quite big and it involves a lot of changes.

File: librz/include/rz_core.h [hunk 11]
```diff
@@ -836,8 +859,12 @@ RZ_API int rz_core_disasm_pdi(RzCore *core, int nb_opcodes, int nb_bytes, int fm
...
 RZ_API int rz_core_flag_in_middle(RzCore *core, ut64 at, int oplen, int *midflags);
 RZ_API int rz_core_bb_starts_in_middle(RzCore *core, ut64 at, int oplen);
+RZ_API void rz_analysis_disasm_text_free(RZ_NULLABLE void *p);
+RZ_API bool rz_disasm_check_end(int nb_opcodes, int i_opcodes, int nb_bytes, int i_bytes);
```
Comment (ret2libc) [inline librz/include/rz_core.h]: does this need to be exported? it seems a very specific function with no real place in an API.

File: librz/core/cmd/cmd_print.c [hunk 12]
```diff
@@ -5233,14 +5018,9 @@ RZ_IPI int rz_cmd_print(void *data, const char *input) {
 			}
 			break;
 		}
-		case 'd': // "pId" is the same as pDi
-			if (l) {
-				rz_core_disasm_pdi(core, 0, l, 0);
-			}
-			break;
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:5250]: what is replacing `pDi`?

File: librz/core/cmd/cmd_print.c [hunk 13]
```diff
@@ -5266,16 +5046,6 @@ RZ_IPI int rz_cmd_print(void *data, const char *input) {
...
-		case 'j': // pij is the same as pdj
-			if (l != 0) {
-				cmd_pdj(core, input + 2, block);
-			}
-			break;
-		case 'd': // "pid" is the same as pdi
-			if (l != 0) {
-				rz_core_disasm_pdi(core, l, 0, 0);
-			}
-			break;
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:5288]: where are these going?

File: librz/core/cmd/cmd_print.c [hunk 14]
```diff
@@ -5418,399 +5188,6 @@ RZ_IPI int rz_cmd_print(void *data, const char *input) {
...
-			bw_disassemble = true;
-			l = use_blocksize; // negative
-			use_blocksize = -use_blocksize;
-		} else {
-			l = use_blocksize;
-		}
-		// may be unnecessary, fixes 'pd 1;pdj 100;pd 1' bug
-		rz_core_block_read(core);
-
-		switch (input[1]) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:5496]: where are all these following commands going?

File: librz/include/rz_core.h [hunk 15]
```diff
@@ -813,6 +822,20 @@ typedef struct rz_core_asm_hit {
...
+/**
+ * \brief Disassemble Options, just for rz_core_print_disasm
+ */
+typedef struct rz_core_disasm_options {
+	int invbreak;
+	int cbytes;
+	bool json; ///< Print in json
+	PJ *pj; ///< PJ instance
+	RzAnalysisFunction *function; ///< Disassemble a function
+	RzList *out_list; ///< Not print, but append as RzList<RzAnalysisDisasmText>
```
Comment (ret2libc) [inline librz/include/rz_core.h]: is there any place where this is set to anything?

**Review by ret2libc — COMMENTED** [review]
Thanks for working on this, it is a very challenging task to get out of this mess the current code is! I left some comments for now, however I feel we could do even better by writing some unit tests that use these functions as that would help understand how we want the APIs to look like.

File: librz/core/disasm.c (hunk 7 above)
Comment (wargio) [reply librz/core/disasm.c]: Add the RZ_NULLABLE anyway. i add it everywhere for the swig apis.

Comment (ret2libc) [reply librz/core/disasm.c]: 👍

File: librz/core/cmd/cmd_print.c (hunk 12 above)
Comment (b1llow) [reply librz/core/cmd/cmd_print.c:5250]: `pId` is an alias for `pDi`, now they are all assigned to `pDq`
Similarly, `pid` | `pdi` -> `pdq`

File: librz/core/cmd/cmd_print.c (hunk 13 above)
Comment (b1llow) [reply librz/core/cmd/cmd_print.c:5288]: https://github.com/rizinorg/rizin/pull/2552
They're newshells now, just some old styles code hasn't been removed

File: librz/core/disasm.c (hunk 8 above)
Comment (b1llow) [reply librz/core/disasm.c]: It's just a different way, I don't think there's anything wrong with me

File: librz/include/rz_core.h (hunk 10 above)
Comment (b1llow) [reply librz/include/rz_core.h]: But there is no problem, if just `aoj` `aOj` `a8` calls `rz_core_analysis_bytes`.

Just breaks the tests when I rewrite `rz_core_print_disasm_json` with `rz_core_analysis_bytes`.

File: librz/core/disasm.c [hunk 16]
```diff
@@ -5905,58 +5982,54 @@ RZ_API int rz_core_print_disasm_instructions_with_buf(RzCore *core, ut64 address
...
 
-static bool handle_backwards_disasm(RzCore *core, int *nb_opcodes, int *nb_bytes) {
-	if (!*nb_opcodes && !*nb_bytes) {
+/**
+ * \brief Handle negative \p pn_opcodes and \p pn_bytes
+ * 	  and seek new offset
+ * 	  and read to core->block
+ * \param core RzCore reference
+ * \param pn_opcodes Pointer to n_opcodes
+ * \param pn_bytes Pointer to n_bytes
```
Comment (ret2libc) [inline librz/core/disasm.c]: could you rewrite this a bit better for someone who does not know exactly what are n_opcodes/n_bytes? What if I provide both? Which one is respected?

File: librz/core/disasm.c [hunk 17]
```diff
@@ -5274,9 +5319,37 @@ static void ds_end_line_highlight(RzDisasmState *ds) {
...
+	if (!p) {
+		return;
+	}
+	RzAnalysisDisasmText *t = p;
+	free(t->text);
+	free(t);
+}
+
+/**
+ * \brief Disassemble \p len bytes and \p nlines opcodes, or funitcon, restricted by \p len and \p nlines at the same time
```
Comment (ret2libc) [inline librz/core/disasm.c]: functions?

does it allow to disassembly X functions?

File: librz/core/disasm.c [hunk 18]
```diff
@@ -1662,6 +1682,31 @@ static int handleMidBB(RzCore *core, RzDisasmState *ds) {
 	return 0;
 }
 
+/**
+ * \brief Update \p oplen by "asm.bb.middle" and "asm.flags.middle"
```
Comment (ret2libc) [inline librz/core/disasm.c:1686]: Could this be rewritten a bit better? It's not clear to me what it does.

File: librz/core/disasm.c [hunk 19]
```diff
@@ -1662,6 +1682,31 @@ static int handleMidBB(RzCore *core, RzDisasmState *ds) {
...
+ * \param core RzCore reference
+ * \param at Address
+ * \param[in,out] oplen Opcode length
+ * \param ret Value by call rz_asm_disassemble
```
Comment (ret2libc) [inline librz/core/disasm.c:1690]: ? 

I wonder whether this really needs to be a RZ_API

**Review by ret2libc — CHANGES_REQUESTED** [review]
Blocking for now.

File: librz/core/disasm.c (hunk 18 above)
Comment (b1llow) [reply librz/core/disasm.c:1686]: I also don't clear what it does, and it wasn't commented before

Comment (b1llow) [reply librz/core/disasm.c:1686]: https://github.com/rizinorg/rizin/blob/5f826df878145793d7d6cca091a2a44390ffd76a/librz/core/disasm.c#L5791-L5801

File: librz/core/cmd/cmd_print.c (hunk 13 above)
Comment (XVilka) [reply librz/core/cmd/cmd_print.c:5288]: I think it's fine then 👍

File: librz/core/cmd/cmd_print.c (hunk 14 above)
Comment (XVilka) [reply librz/core/cmd/cmd_print.c:5496]: They were ported to the rzshell https://github.com/rizinorg/rizin/pull/2552

**Review by XVilka — APPROVED** [review]
Looks much better!

**Comment by ret2libc** [comment]
Please work on adding at least 1 unit test if you can. That will help make sure the API makes sense and kinda works.

File: librz/include/rz_core.h [hunk 20]
```diff
@@ -434,6 +434,15 @@ typedef struct rz_bin_sym_names_t {
...
 } RzBinSymNames;
 
+/**
+ * \brief Message for `pdJ`
+ */
+typedef struct rz_analysis_disasm_text_t {
+	ut64 offset;
+	ut64 arrow;
+	char *text;
+} RzAnalysisDisasmText;
```
Comment (ret2libc) [inline librz/include/rz_core.h]: Is this really just for `pdJ` command? If so, is it really necessary?

Comment (b1llow) [reply librz/include/rz_core.h]: Yes

**Comment by XVilka** [comment]
Agree on unit tests for these functions - they are one of a few most important functions in the whole of Rizin, and we should be careful about them.

File: librz/core/disasm.c [hunk 21]
```diff
@@ -5292,17 +5365,19 @@ RZ_API int rz_core_print_disasm(RzPrint *p, RzCore *core, ut64 addr, ut8 *buf, i
...
+	ds->l = nlines;
 	ds->buf = buf;
 	ds->len = len;
 	ds->addr = addr;
 	ds->hint = NULL;
 	ds->buf_line_begin = 0;
-	ds->pdf = pdf;
+	ds->pdf = options ? options->function : NULL;
+	ds->pj = NULL;
+	ds->vec = options ? options->vec : NULL;
```
Comment (ret2libc) [inline librz/core/disasm.c:5378]: Again, I think this is not alright. `ds->vec` should be initialized with a new Rzvector imho because you are asking `rz_core_print_disasm` to disassemble some instructions and the result will be in `->vec`, thus the vec will be filled from the function and I don't see reasons to append stuff to existing instructions.

File: librz/core/disasm.c [hunk 22]
```diff
@@ -5274,9 +5319,37 @@ static void ds_end_line_highlight(RzDisasmState *ds) {
...
+ * \param options Disassemble Options
+ * \return Disassemble bytes number
+ */
+RZ_API int rz_core_print_disasm(RZ_NONNULL RzCore *core, ut64 addr, RZ_NONNULL ut8 *buf, int len, int nlines,
+	RZ_NULLABLE RzCoreDisasmOptions *options) {
+	rz_return_val_if_fail(core && buf, 0);
+
+	PJ *pj = options ? options->pj : NULL;
+	bool json = options ? options->json : false;
+
```
Comment (ret2libc) [inline librz/core/disasm.c]: Did you consider using `RzCmdStateOutput`?

Comment (ret2libc) [reply librz/core/disasm.c]: Also, pass that as argument to this function instead of part of `RzCoreDisasmOptions`

File: librz/core/disasm.c [hunk 23]
```diff
@@ -5905,58 +5982,66 @@ RZ_API int rz_core_print_disasm_instructions_with_buf(RzCore *core, ut64 address
...
+ * 	nbytes = nbytes ? min(abs(nbytes), RZ_CORE_MAX_DISASM)
+ * 			: max(core->blocksize, min(abs(nbytes), RZ_CORE_MAX_DISASM))
+ * 	offset = rz_core_prevop_addr_force(...)
+ * } else {
+ * 	offset = core->offset + (nbytes>=0 ? 0 : -nbytes)
+ * 	nopcodes = nopcodes
+ *	nbytes = nbytes ? min(abs(nbytes), RZ_CORE_MAX_DISASM)
+ *			: max(core->blocksize, min(abs(nbytes), RZ_CORE_MAX_DISASM))
+ * }
+ *
```
Comment (ret2libc) [inline librz/core/disasm.c]: Purpose of documentation is that i want to know how to use an API without reading its implementation, having this code does not help IMHO.

File: test/db/cmd/cmd_print [hunk 24]
```diff
@@ -425,36 +425,37 @@ wx b8010000004839ca7f00
 pi -3 @ 10
 EOF
 EXPECT=<<EOF
-add byte [rax], al
+mov eax, 1
```
Comment (ret2libc) [inline test/db/cmd/cmd_print:428]: awesome! 👍

File: librz/core/canalysis.c [hunk 25]
```diff
@@ -6809,10 +6814,6 @@ RZ_API void rz_analysis_bytes_free(RZ_NULLABLE void *ptr) {
  */
 RZ_API RZ_OWN RzPVector *rz_core_analysis_bytes(RZ_NONNULL RzCore *core, RZ_NONNULL const ut8 *buf, int len, int nops) {
```
Comment (ret2libc) [inline librz/core/canalysis.c:6817]: Do you know why `cmd_analysis.c:core_analysis_bytes_json()` uses this function but `core_analysis_bytes_json:core_analysis_bytes_standard()` does not? Would it make sense to use this function also there to collect the information about the analyzed bytes and iterate over the result vec?

File: librz/core/cmd/cmd_print.c [hunk 26]
```diff
@@ -5418,399 +5188,6 @@ RZ_IPI int rz_cmd_print(void *data, const char *input) {
...
-					rz_core_block_size(core, bsz);
-				}
-			}
-			l = 0;
-			break;
-		case 'p': // "pdp"
-			processed_cmd = true;
-			disasm_ropchain(core, core->offset, 'D');
-			pd_result = true;
-			break;
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:5682]: is this moved elsewhere?

**Review by ret2libc — COMMENTED** [review]
I tested a bit the PR and looks nice so far! I think it's just a matter of making sure some parts of the code are clearer/cleaner for the future, but very good job overall! `pd`/`pD` are now in very good shape!

File: librz/core/cmd/cmd_print.c (hunk 26 above)
Comment (b1llow) [reply librz/core/cmd/cmd_print.c:5682]: Yes, it's handled by `rz_cmd_disassemble_ropchain_handler` now.

File: librz/core/disasm.c (hunk 21 above)
Comment (b1llow) [reply librz/core/disasm.c:5378]: Just because our cutter uses the `pdJ` command,
So I want to get this data on the caller side instead of a string in JSON format.

```
[0x00401cf0]> pdJ 3 | jq
[
  {
    "offset": 4201712,
    "text": "        \u001b[34m╎\u001b[0m   \u001b[36m;-- entry0:\u001b[0m"
  },
  {
    "offset": 4201712,
    "text": "        \u001b[34m╎\u001b[0m   \u001b[32m0x00401cf0\u001b[0m      \u001b[1;92mcall  0x401fdb\u001b[0m\u001b[0m"
  },
  {
    "offset": 4201717,
    "arrow": 4201161,
    "text": "        \u001b[34m└\u001b[0m\u001b[34m─\u001b[0m\u001b[34m<\u001b[0m \u001b[32m0x00401cf5\u001b[0m      \u001b[32mjmp   0x401ac9\u001b[0m\u001b[0m"
  },
  {
    "offset": 4201722,
    "text": "            \u001b[32m0x00401cfa\u001b[0m      \u001b[1;91mint3\u001b[0m\u001b[0m\u001b[0m"
  }
]
```

Comment (ret2libc) [reply librz/core/disasm.c:5378]: Yes, I agree, but it should be something like
```
ds->vec = rz_list_newf(....);
```
instead of what it is now.

Comment (b1llow) [reply librz/core/disasm.c:5378]: But if I do this, I can only get this data on the caller by using `ds->vec` as the return value, and then this breaks the current return value system.

And now, I just need

```c
RzPVector* vec = rz_pvector_new(...)
rz_core_print_disasm(...vec)
```

If I do not need the data

```c
rz_core_print_disasm(...vec:NULL)
```

**Comment by b1llow** [comment]
```
../subprojects/libzip-1.7.3/lib/zip_fdopen.c: In function ‘zip_fdopen’:
../subprojects/libzip-1.7.3/lib/zip_fdopen.c:62:15: error: implicit declaration of function ‘fdopen’; did you mean ‘fopen’? [-Werror=implicit-function-declaration]
   62 |     if ((fp = fdopen(fd, "rb")) == NULL) {
      |               ^~~~~~
      |               fopen
../subprojects/libzip-1.7.3/lib/zip_fdopen.c:62:13: error: assignment to ‘FILE *’ {aka ‘struct _IO_FILE *’} from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
   62 |     if ((fp = fdopen(fd, "rb")) == NULL) {
      |             ^
```

**Comment by XVilka** [comment]
@imbillow something has changed in the GitHub Actions, I get the same error on other PRs: https://github.com/rizinorg/rizin/runs/7172329650?check_suite_focus=true#step:11:948

**Comment by XVilka** [comment]
Another possible reason is the new version of Meson:
- https://github.com/mesonbuild/meson/releases/tag/0.63.0
- https://mesonbuild.com/Release-notes-for-0-63-0.html

I think, it's the cause of everything broken suddenly.

File: test/unit/test_analysis_op.c [hunk 27]
```diff
@@ -96,8 +97,63 @@ bool test_rz_analysis_op_val() {
...
+	mu_assert_streq(ab->pseudo, "dword [rbp - 4] = edi", "rz_core_analysis_bytes pseudo");
+
+	rz_core_free(core);
+	rz_pvector_free(vec);
+	mu_end;
+}
+
+bool test_rz_core_print_disasm() {
+	RzCore *core = rz_core_new();
+	ut8 buf[128];
```
Comment (ret2libc) [inline test/unit/test_analysis_op.c]: SWITCH_TO_ARCH_BITS as well here?

File: librz/core/disasm.c [hunk 28]
```diff
@@ -5274,9 +5319,37 @@ static void ds_end_line_highlight(RzDisasmState *ds) {
...
+RZ_API void rz_analysis_disasm_text_free(RzAnalysisDisasmText *t) {
+	if (!t) {
+		return;
+	}
+	free(t->text);
+	free(t);
+}
+
+/**
+ * \brief Disassemble \p len bytes and \p nlines opcodes, or a funitcon,
```
Comment (ret2libc) [inline librz/core/disasm.c]: `or a function`? I would remove that part.

**Review by XVilka — APPROVED** [review]
I think it's in a really good shape. Of course, there are some improvements could be made in this code, but it can be done later, in a separate PRs. I propose merging this as is. cc @ret2libc @wargio @thestr4ng3r

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
Why are there command name changes? This and the overall goal should be elaborated in the description.

**Review by thestr4ng3r — APPROVED** [review]

**Comment by thestr4ng3r** [comment]
I would still wait for @ret2libc to approve

File: librz/core/canalysis.c [hunk 29]
```diff
@@ -6821,41 +6824,69 @@ RZ_API RZ_OWN RzPVector *rz_core_analysis_bytes(RZ_NONNULL RzCore *core, RZ_NONN
...
-	int ret;
-	for (int i = 0, idx = 0; idx < len && (!nops || (nops && i < nops)); i++, idx += ret) {
+	bool bigendian = rz_config_get_b(core->config, "cfg.bigendian");
+	bool asm_sub_var = rz_config_get_i(core->config, "asm.sub.var");
+	core->parser->subrel = rz_config_get_i(core->config, "asm.sub.rel");
+	core->parser->localvar_only = rz_config_get_i(core->config, "asm.sub.varonly");
+
+	const int addrbytes = (int)core->io->addrbytes;
+	const int mask = RZ_ANALYSIS_OP_MASK_ESIL | RZ_ANALYSIS_OP_MASK_IL | RZ_ANALYSIS_OP_MASK_OPEX | RZ_ANALYSIS_OP_MASK_HINT;
+	RzAsmOp asmop;
```
Comment (wargio) [inline librz/core/canalysis.c:6834]: This value is never initialized. rz_asm_op_init should be called before being used.

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/core/canalysis.c (hunk 29 above)
Comment (b1llow) [reply librz/core/canalysis.c:6834]: https://github.com/rizinorg/rizin/blob/628d04f713bb3fdcd5726f00d4f440ead35dd683/librz/asm/asm.c#L462-L464
but `rz_asm_disassemble` will do this, so is this necessary?

**Review by wargio — APPROVED** [review]

File: test/unit/test_analysis_op.c [hunk 30]
```diff
@@ -96,8 +97,63 @@ bool test_rz_analysis_op_val() {
...
+
+	mu_assert_eq(rz_pvector_len(vec), 3, "rz_core_print_disasm len");
+	RzAnalysisDisasmText *t = rz_pvector_at(vec, 0);
+	mu_assert_eq(t->offset, 0, "rz_core_print_disasm offset");
+
+	t = rz_pvector_at(vec, 1);
+	mu_assert_eq(t->offset, 1, "rz_core_print_disasm offset");
+
+	t = rz_pvector_at(vec, 2);
+	mu_assert_eq(t->offset, 4, "rz_core_print_disasm offset");
```
Comment (ret2libc) [inline test/unit/test_analysis_op.c]: check also the ->text and ->arrow. Also, please add some doxygen doc for the ->arrow field because it is not clear what it represents.

File: librz/include/rz_core.h [hunk 31]
```diff
@@ -811,6 +820,18 @@ typedef struct rz_core_asm_hit {
...
 } RzCoreAsmHit;
 
+/**
+ * \brief Disassemble Options, just for rz_core_print_disasm
+ */
+typedef struct rz_core_disasm_options {
+	int invbreak;
+	int cbytes;
+	RzAnalysisFunction *function; ///< Disassemble a function
+	RzPVector *vec; ///< Not print, but append as RzList<RzAnalysisDisasmText>
```
Comment (XVilka) [inline librz/include/rz_core.h]: I guess it should be `RzPVector<RzAnalysisDisasmText>`

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]
